File: app/src/main/java/com/stwpower/powertap/utils/LoggingInterceptor.java
package com.stwpower.powertap.utils;

import android.util.Log;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import okio.Buffer;

public class LoggingInterceptor implements Interceptor {
    private static final String TAG = "网络请求";

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request request = chain.request();

        long t1 = System.nanoTime();
        Log.d(TAG, String.format("Sending request %s on %s%n%s",
                request.url(), chain.connection(), request.headers()));

        if (request.method().compareToIgnoreCase("post")==0) {
            Request copy = request.newBuilder().build();
            Buffer buffer = new Buffer();
            if(copy.body() != null)
                copy.body().writeTo(buffer);
            Log.d(TAG, "Request body: " + buffer.readUtf8());
        }

        Response response = chain.proceed(request);

        long t2 = System.nanoTime();
        String responseBody = response.peekBody(Long.MAX_VALUE).string();
        Log.d(TAG, String.format("Received response for %s in %.1fms%n%s",
                response.request().url(), (t2 - t1) / 1e6d, response.headers()));
        Log.d(TAG, "Response body: " + responseBody);
        return response;
    }
}

File: app/src/main/java/com/stwpower/powertap/utils/LoggingInterceptor.kt
package com.stwpower.powertap.utils

import android.util.Log

import java.io.IOException

import okhttp3.Interceptor
import okhttp3.Response
import okio.Buffer

class LoggingInterceptor : Interceptor {
    companion object {
        private const val TAG = "powertap"
    }

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()

        val t1 = System.nanoTime()
        MyLog.d("Sending request ${request.url} on ${chain.connection()}\n${request.headers}")

        // 打印 POST 请求 body
        if (request.method.equals("POST", ignoreCase = true)) {
            val copy = request.newBuilder().build()
            val buffer = Buffer()
            copy.body?.writeTo(buffer)
            MyLog.d("Request body: ${buffer.readUtf8()}")
        }

        val response = chain.proceed(request)

        val t2 = System.nanoTime()
        val responseBody = response.peekBody(Long.MAX_VALUE).string()
        MyLog.d("Received response for ${response.request.url} in ${(t2 - t1) / 1e6}ms\n${response.headers}")
        MyLog.d("Response body: $responseBody")

        return response
    }
}
